Reimplement ColorPicker presets

Geometror opened the pull request:
Reimplements the color preset functionality of the ColorPicker:

![grafik](https://user-images.githubusercontent.com/50084500/129491085-e121ac1e-76d0-46f4-ad83-82bcb3408520.png)

Changes:
-	Reimplement color picker presets with custom Button (ColorPresetButton)
Presets were disabled due to the way they were implemented as stated by reduz's comment: `//presets should be shown using buttons or something else, this method is not a good idea` (the former method was using a TextureRect for the whole palette)
-	Add the overbright indicator which is already used in other places of the ColorPicker and an alpha checker board background for transparent colors
-	Some small structural improvements to the code for better readability

fire commented:
This design is on the right track. Something came up, maybe @groud can review.

pycbouh commented on scene/gui/color_picker.cpp:
```diff
@@ -422,14 +416,39 @@ ColorPicker::PickerShapeType ColorPicker::get_picker_shape() const {
+		// Find button to move to the end.
+		ColorPresetButton *btn_to_move = nullptr;
+		for (int i = 1; i < preset_container->get_child_count(); i++) {
+			ColorPresetButton *current_btn = (ColorPresetButton *)preset_container->get_child(i);
```
You should probably use `Object::cast_to<ColorPresetButton>` here, and make sure that the cast was successful.

Geometror replied:
Done.

Chaosus commented on scene/gui/color_picker.cpp:
```diff
@@ -932,42 +963,18 @@ void ColorPicker::_w_input(const Ref<InputEvent> &p_event) {
 }
 
-void ColorPicker::_preset_input(const Ref<InputEvent> &p_event) {
+void ColorPicker::_preset_input(const Ref<InputEvent> &p_event, Color p_color) {
```
Change it to `const Color& p_color`

Geometror replied:
Fixed. Changed the color parameter to be passed as a const reference also for the add_preset_button(...) function.

Chaosus commented on scene/gui/color_picker.h:
```diff
@@ -119,14 +138,16 @@ class ColorPicker : public BoxContainer {
 	void _uv_input(const Ref<InputEvent> &p_event, Control *c);
 	void _w_input(const Ref<InputEvent> &p_event);
-	void _preset_input(const Ref<InputEvent> &p_event);
+	void _preset_input(const Ref<InputEvent> &p_event, Color p_button);
```
`Color p_button` -> `const Color& p_color` or `p_button_color`

Geometror replied:
Oops, thought I changed this. Fixed.

KoBeWi commented:
I get this crash when I try to remove preset:
```
CrashHandlerException: Program crashed
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] StyleBoxFlat::set_bg_color (C:\godot_source\scene\resources\style_box.cpp:315)
[1] ColorPresetButton::_notification (C:\godot_source\scene\gui\color_picker.cpp:1452)
[2] ColorPresetButton::_notificationv (C:\godot_source\scene\gui\color_picker.h:48)
[3] Object::notification (C:\godot_source\core\object\object.cpp:843)
[4] CanvasItem::_update_callback (C:\godot_source\scene\main\canvas_item.cpp:153)
[5] call_with_variant_args_helper<CanvasItem> (C:\godot_source\core\variant\binder_common.h:227)
[6] call_with_variant_args_dv<CanvasItem> (C:\godot_source\core\variant\binder_common.h:370)
[7] MethodBindT<CanvasItem>::call (C:\godot_source\core\object\method_bind.h:287)
[8] Object::call (C:\godot_source\core\object\object.cpp:832)
[9] Callable::call (C:\godot_source\core\variant\callable.cpp:62)
[10] MessageQueue::_call_function (C:\godot_source\core\object\message_queue.cpp:259)
[11] MessageQueue::flush (C:\godot_source\core\object\message_queue.cpp:306)
[12] Main::iteration (C:\godot_source\main\main.cpp:2521)
[13] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:626)
[14] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:163)
[15] _main (C:\godot_source\platform\windows\godot_windows.cpp:185)
[16] main (C:\godot_source\platform\windows\godot_windows.cpp:199)
[17] __scrt_common_main_seh (D:\a01\_work\26\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[18] BaseThreadInitThunk
-- END OF BACKTRACE --
```

EricEzaM commented:
Looks like the destroctor runs which calls unref on a static property, so the next time draw is called on any button it crashes.

KoBeWi commented:
Is there even a need for cached stylebox? They aren't very expensive. You could create them on demand to prevent having lots of unused styleboxes in every picker. Or maybe put it in theme and add checks that ensure it's StyleboxFlat (or a StyleboxTexture, which you could colorize; that would be interesting option for customization).

KoBeWi commented on scene/gui/color_picker.h:
```diff
@@ -83,14 +98,18 @@ class ColorPicker : public BoxContainer {
-	int presets_per_row = 0;
 	PickerShapeType picker_type = SHAPE_HSV_WHEEL;
 
+	int preset_column_count = 9;
```
```suggestion
	const int preset_column_count = 9;
```
It could be constant, it's not modified anywhere.

Geometror replied:
Done.

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -422,14 +416,39 @@ ColorPicker::PickerShapeType ColorPicker::get_picker_shape() const {
 		presets.push_back(p_color);
 		preset_cache.push_back(p_color);
+
+		int hsep = preset_container->get_theme_constant(SNAME("hseparation"));
+		int preset_size = (int(get_size().width) - (hsep * (preset_column_count - 1))) / preset_column_count;
```
This snippet appears twice in the code. You could make a (inline?) function to avoid code duplication, e.g.
```
inline int _get_preset_size() { return (int(get_size().width) - (preset_container->get_theme_constant(SNAME("hseparation")) * (preset_column_count - 1))) / preset_column_count; }
```

Geometror replied:
Done.

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -422,14 +416,39 @@ ColorPicker::PickerShapeType ColorPicker::get_picker_shape() const {
+				break;
+			}
+		}
+		if (btn_to_move) {
+			preset_container->move_child(btn_to_move, preset_container->get_child_count() - 1);
+		}
```
You break the loop when you find button, so you could just move the button inside loop instead and then `btn_to_move` variable is not necessary. Same with removing few lines below.

Geometror replied:
Done.

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -1428,3 +1431,55 @@ void ColorPickerButton::_bind_methods() {
+
+void ColorPresetButton::set_preset_color(const Color &p_color) {
+	preset_color = p_color;
+	//stylebox->set_bg_color(p_color);
```
Commented out code. Can be removed.

Geometror replied:
Needed again.

KoBeWi commented:
You could actually avoid adding a new class if you made a flat button and used `draw` signal to handle custom drawing inside ColorPicker.

Geometror commented:
Fixed the crash. (the cached stylebox was a bad last-minute decision without testing :/)
I chose to implement the presets with a custom button class because a normal button is more expensive to create (even with no text a shaped_text gets created) and it felt less hacky in general. But if adding a new class is undesired, I will change that.
The idea to make the color preset stylebox part of the theme is pretty good, I will look into that tomorrow :)

KoBeWi commented:
Ah, you are right. Having a custom button is more light-weight in this case, so it's ok.

Geometror commented:
- Added theming support for the ColorPresetButton (StyleBoxFlat and StyleBoxTexture are supported)
	- the foreground can be changed (StyleBoxFlat and StyleBoxTexture are supported)
	- the overbright indicator can be changed (default is white corner triangle)
	- the preset background texture can be changed (default is MiniCheckerboard)
Now something like this is possible when using a ColorPicker in a project:
![grafik](https://user-images.githubusercontent.com/50084500/129784551-3e4df45e-0745-47c3-a5d0-b758fa2d2d47.png)
- Moved splitted ColorPickerButton theming in default_theme.cpp (there was one large codeblock and one line at another place in the file) Now ColorPicker, ColorPresetButton and ColorPickerButton are all at one place.
- Renamed preset_bg to sample_bg to be more descriptive (because it was only used for the sample area in the ColorPicker)/ preset_bg is now actually used for the presets
- Optimization: only update the presets when the their desired size has changed during a theme change (NOTIFICATION_THEME_CHANGED is still sent 4-8 times when selecting another node in the scene tree)

pycbouh commented:
With styling support like this I kind of think it could be made into a separate `ColorPalette` control, so it could be used independent from the `ColorPicker`.

KoBeWi commented:
I think that could be done in another PR.

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -1428,3 +1431,64 @@ void ColorPickerButton::_bind_methods() {
+			const Rect2 r = Rect2(Point2(0, 0), get_size());
+			Ref<StyleBox> sb_raw = get_theme_stylebox(SNAME("preset_fg"), SNAME("ColorPresetButton"))->duplicate();
+
+			if (sb_raw->get_class_name() == SNAME("StyleBoxFlat")) {
```
StyleBoxes are Resources, so you can assign-check them like InputEvents (see any input method), e.g.
```
Ref<StyleBoxFlat> sb_flat = sb_raw;
if (sb_raw.is_valid()) {
    ...
}
```

Geometror replied:
Done.

KoBeWi reviewed (approved):
Looks ok overall, ~~but change the conditions in `NOTIFICATION_DRAW` like I commented.~~

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -1428,3 +1431,64 @@ void ColorPickerButton::_bind_methods() {
+				sb_texture->set_modulate(preset_color);
+				sb_texture->draw(get_canvas_item(), r);
+			} else {
+				WARN_PRINT("Unsupported StyleBox used for ColorPresetButton. Use StyleBoxFlat and StyleBoxTexture instead.");
```
```suggestion
				WARN_PRINT("Unsupported StyleBox used for ColorPresetButton. Use StyleBoxFlat or StyleBoxTexture instead.");
```
Spotted just now...

Geometror replied:
Fixed.

Geometror committed:
Reimplement ColorPicker presets

fire commented:
Thanks.

akien-mga commented:
This would need a dedicated backport PR for `3.x` if wanted there. Maybe also without the theme item rename to avoid breaking compat? Provided that it was working in 3.3.
